### ISSUE TITLE
Generics don't work

### ISSUE THREAD

**Issue by programWhiz** [vlang/v#8365]
V full version: V 0.2.2 29d6e40.1f4e825
OS: macos, Mac OS X, 10.15.7, 19H2
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
CC version: Apple clang version 12.0.0 (clang-1200.0.32.28)

**EDIT** See below, I believe the issue is that the parser does not correctly propagate the template info from the return type, and it fails to propagate this to the struct fields in constructor.

Tried simple example with some generics, it crashes in compiler trying to access array out of bounds.  My guess is that the list of generics isn't properly populated so when it tries to find `T` in the list, its crashing.  It will crash earlier if I declare it with "member function" style and put `(p Pool<T>)` prior to function name.  It seems to go further without crsahing if I use the `fn F<T>(p Pool<T>, ...)` style.

Minimal reproducing program:
```v
struct Foo {
    id u64
}

pub struct Pool<T> {
pub mut:
    objs [] T
}

pub fn get_pool_cap<T>(p Pool<T>) int {
    return p.objs.cap
}

fn create_pool<T>( cap int ) Pool<T> {
    return Pool<T> {
        objs : [] T { cap: cap }
    }
}

fn main() {
    p := create_pool<Foo>(100)
    get_pool_cap<Foo>(p)
}
```

And the error:
```
V panic: array.get: index out of range (i == 0, a.len == 0)
0   v                                   0x000000010a87aa9f array_get + 95
1   v                                   0x000000010a95a4a1 v__gen__Gen_unwrap_generic + 257
2   v                                   0x000000010a96a583 v__gen__Gen_cc_type2 + 259
3   v                                   0x000000010a969dae v__gen__Gen_cc_type + 78
4   v                                   0x000000010a969935 v__gen__Gen_base_type + 69
5   v                                   0x000000010a94e06b v__gen__Gen_typ + 27
6   v                                   0x000000010a96fe4c v__gen__Gen_gen_assign_stmt + 7964
7   v                                   0x000000010a96b319 v__gen__Gen_stmt + 265
8   v                                   0x000000010a96af3b v__gen__Gen_stmts_with_tmp_var + 363
9   v                                   0x000000010a963bf3 v__gen__Gen_stmts + 83
10  v                                   0x000000010a975eaf v__gen__Gen_gen_fn_decl + 7727
11  v                                   0x000000010a96c270 v__gen__Gen_stmt + 4192
12  v                                   0x000000010a96af3b v__gen__Gen_stmts_with_tmp_var + 363
13  v                                   0x000000010a963bf3 v__gen__Gen_stmts + 83
14  v                                   0x000000010a961f50 v__gen__cgen + 5520
15  v                                   0x000000010a9bbee5 v__builder__Builder_gen_c + 549
16  v                                   0x000000010a9bc5b7 v__builder__Builder_build_c + 295
17  v                                   0x000000010a9bca0e v__builder__Builder_compile_c + 846
18  v                                   0x000000010a9c79e3 v__builder__compile + 339
19  v                                   0x000000010a9cc4e9 main__main + 4713
20  v                                   0x000000010a9cebaa main + 42
21  v                                   0x000000010a876694 start + 52
22  ???                                 0x0000000000000005 0x0 + 5
```

**Comment by programWhiz** [comment]
Additional info.  This is triggered by the struct definition on line 625 of cgen.v (not by multi-return).  Additionally, the error is on line 329 of fn.v:
```
	return g.cur_generic_types[idx].derive(typ).clear_flag(.generic)
```

The `idx` is zero because there are no types in the loop.  I added this logic to see what would happen if we continued as normal:
```
mut idx := -1
...
if idx < 0 {
   return typ
}
return g.cur_generic_types[idx].derive(typ).clear_flag(.generic)
```

This allowed the parser to at least generate the C program, but the issue is that the struct template instantiation isn't defined.  I get to the line where the functions are being declared, but the template specialization isn't declared in the .c file:
```
// These are present
typedef struct main__Foo main__Foo;
typedef struct main__Pool main__Pool;

// but these functions require the specialized versions
int main__get_pool_cap_T_main__Foo(main__Pool_T_main__Foo p);
VV_LOCAL_SYMBOL main__Pool_T_main__Foo main__create_pool_T_main__Foo(int cap);
```

**Comment by programWhiz** [comment]
To make sure the type was present in the .c file, I went ahead and forward-declared the type in the `test.v` file I pasted in the original bug report:
```type FooPool = Pool<Foo>```

This allowed the parser to create `main__Pool_T_main__Foo`.  However, there issues now because the compile thinks the return value from the `create_pool` function is a `main__Pool` (no template), but recognizes that the params to `get_pool_cap` should be `main__Pool__T_main__Foo`.   I changed to pointers vs raw structs, but it still had the same issue:
```
incompatible pointer types initializing 'main__Pool *' (aka 'struct main__Pool *') with an expression of type 'main__Pool_T_main__Foo *' (aka 'struct main__Pool_T_main__Foo *')
```

This leads me to believe the bug is actually the fact that the template specialization needs to occur on the return type specialization.

One last experiment, I went ahead and inlined the creation function, but this had a similar failure:
```
fn main() {
    p := &Pool<Foo> { }
    get_pool_cap<Foo>(p)
}
```

The failure is now on allocating values for `Pool.objs`, it seems the template system didn't know to propagate the template down to objs:
```
use of undeclared identifier 'T'
        main__Pool_T_main__Foo* p = (main__Pool_T_main__Foo*)memdup(&(main__Pool_T_main__Foo){.objs = __new_array(0, 1, sizeof(T)),}, sizeof(main__Pool_T_main__Foo));
```

If I try to specifially set it with `objs: [] Foo` then v will complain: 
```
cannot assign to field `objs`: expected `[]T`, not `[]Foo`
```

**Comment by danieldaeschle** [comment]
Thank you for this report. Generics are still WIP. Every issue helps me improving it. Keep testing and reporting. Thanks! 👍🏼

**Comment by danieldaeschle** [comment]
Problem: `resolve_generic_type` in `check_types` doesn't resolve Structs. But since structs generics work different, this needs more investigation.

@joe-conigliaro

**Comment by KeepZen** [comment]
```v
struct Num<T>{
	num T
}
fn (num Num<T>) is_autom() bool{
	return true
}
fn main(){
	num :=Num<int>{3}
	println(num.is_autom())
}
```
Will yield 
  > error: unknown method: `Num<int>.is_autom`.

**Comment by Terisback** [comment]
This is fixed via #9469 @danieldaeschle 
![image](https://user-images.githubusercontent.com/26527529/112596122-33489780-8e2d-11eb-889a-7c60afd6fba4.png)

**Comment by Terisback** [comment]
> ```v
> struct Num<T>{
> 	num T
> }
> fn (num Num<T>) is_autom() bool{
> 	return true
> }
> fn main(){
> 	num :=Num<int>{3}
> 	println(num.is_autom())
> }
> ```
> 
> Will yield
> 
> > error: unknown method: `Num<int>.is_autom`.

This needs to be made into a separate issue, imo
